### PR TITLE
TMA-732: fix edge cases for user input sanitized MUFs

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 1]
```diff
@@ -143,4 +143,36 @@
       end
     end
   end
+
+  context 'when using sync_multiple_projects_based_on_custom_id mode' do
+    let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:79, synchronize_user_filters_spec.rb:148 (mass = 52)

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 2]
```diff
@@ -143,4 +143,36 @@
...
+        filters_config: { labels: [] },
+        multiple_projects_column: 'id_column',
+        sync_mode: 'sync_multiple_projects_based_on_custom_id',
+        gdc_logger: logger
+      }
+      GoodData::LCM2.convert_to_smart_hash(params)
+    end
+    let(:CSV) { double('CSV') }
+
+    before do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Identical code found in synchronize_user_filters_spec.rb:93, synchronize_user_filters_spec.rb:162 (mass = 152)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -139,6 +139,7 @@ def call(params)
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               filters << row.to_hash
             end
+            fail 'The filter set can not be empty when using sync_multiple_projects_based_on_custom_id mode' if filters.empty?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.empty?' 2 times (DuplicateMethodCall)

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 4]
```diff
@@ -143,4 +143,36 @@
...
+      allow(project).to receive(:uri).and_return('project-uri')
+      allow(project).to receive(:add_data_permissions)
+      allow(domain).to receive(:clients).and_return([])
+      allow(data_source).to receive(:realize).and_return('filepath')
+    end
+
+    it 'fails if the MUF set is empty' do
+      expect(File).to receive(:open)
+      expect(CSV).to receive(:foreach)
+      expect { subject.class.call(params) }.to raise_error /The filter set can not be empty/
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Lint/AmbiguousRegexpLiteral: Ambiguous regexp literal. Parenthesize the method arguments if it's surely a regexp literal, or add a whitespace to the right of the `/` if it should be a division.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/656/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/451/ : FAILURE in 2m 03s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/759/ : SUCCESS in 1m 52s

**Comment by panjan** [comment]
ok to merge

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 5]
```diff
@@ -143,4 +147,26 @@
       end
     end
   end
+
+  context 'when using sync_multiple_projects_based_on_custom_id mode' do
+    include_context 'using mode with custom_id'
+    let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:94, synchronize_user_filters_spec.rb:153 (mass = 52)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/657/ : SUCCESS in 18s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/452/ : FAILURE in 2m 09s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/760/ : SUCCESS in 2m 29s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/658/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/453/ : SUCCESS in 2m 14s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/761/ : SUCCESS in 1m 47s

**Comment by panjan** [comment]
ok to merge

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 6]
```diff
@@ -538,13 +538,16 @@ def self.normalize_filters(filters)
     end
 
     # Removes MUFs from to_delete unless in user is in users_brick_input
+    # if this does not happen, users that are about to be deleted by users_brick
+    # would have all their filters removed now, which is not desirable
     def self.sanitize_filters_to_delete(to_delete, users_brick_input, project_users)
       return to_delete unless users_brick_input && users_brick_input.any?
       user_profiles = users_brick_input.map do |user|
         result = project_users.find { |u| u.login == user['login'] }
```
Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Calls 'filter[:label]' 4 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Contains iterators nested 3 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Has the variable name 'l' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/255/ : SUCCESS in 17s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/459/ : SUCCESS in 1m 58s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/767/ : SUCCESS in 33s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1171/ : FAILURE in 49m 05s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/929/ : SUCCESS in 31m 55s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/264/ : SUCCESS in 18s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/508/ : SUCCESS in 2m 05s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/794/ : SUCCESS in 2m 12s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1182/ : SUCCESS in 54m 35s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/940/ : SUCCESS in 37m 26s

**Pull request merged**